src/tabmat/categorical_matrix.py

```diff
@@ -321,13 +313,21 @@ def __init__(
             )
 
         if not hasattr(cat_vec, "dtype"):
-            cat_vec = np.array(cat_vec)  # avoid errors in pd.factorize
+            cat_vec = np.asarray(cat_vec)  # avoid errors in pd.factorize
 
         self._input_dtype = cat_vec.dtype
         self._missing_method = cat_missing_method
         self._missing_category = cat_missing_name
 
-        indices, self.categories = _extract_codes_and_categories(cat_vec)
+        if categories is not None:
+            self.categories = categories
+            indices = np.nan_to_num(cat_vec, nan=-1)
+            if max(indices) >= len(categories):
+                raise ValueError("Indices exceed length of categories.")
+            if min(indices) < -1:
+                raise ValueError("Indices must be non-negative (or -1 for missing).")
```

Comment:
Would it make sense to check that the indices are integers at this point?

Reply:
Absolutely.

Reply:
I added a try/except block surrounding the call to `.astype(np.int32)`. I don't want to check that the input dtype is an integer directly because I want to allow nan values.